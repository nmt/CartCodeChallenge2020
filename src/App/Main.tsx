import {
  Box,
  Heading,
  Text,
  Strong,
  Card,
  Stack,
} from 'braid-design-system';
import React from 'react';

import { PRICING_RULES, DISCOUNT_STRING, BOGO_STRING, PricingRule } from '../helpers/specialPricingRules';

import GrandTotal from './components/GrandTotal';
import OrderSummary from './components/OrderSummary';
import Profile from './components/Profile';

type state = {
  profile: string,
  items: Array<Item>,
  rules: Array<PricingRule>,
};

interface Item {
  name: string,
  quantity: number,
  price: number,
  subtotal: number,
}

class Main extends React.Component<{}, state> {
  constructor(props: any) {
    super(props);

    this.profileChange = this.profileChange.bind(this);
    this.applyPricingRules = this.applyPricingRules.bind(this);
    this.quantityChange = this.quantityChange.bind(this);

    this.state = {
      profile: '',
      items: [
        {
          name: 'classic',
          quantity: 0,
          price: 26999,
          subtotal: 0,
        },
        {
          name: 'standOut',
          quantity: 0,
          price: 32299,
          subtotal: 0,
        },
        {
          name: 'premium',
          quantity: 0,
          price: 39499,
          subtotal: 0,
        }],
      rules: [],
    };
  }

  applyPricingRules(item: Item, type: string, newQty: number) {
    const rules = this.state.rules;

    // Discount rule
    let relevantRule = rules.find(rule => rule.appliesTo === type && rule.type === DISCOUNT_STRING);
    const newPrice = relevantRule?.details.specialPrice;
    const pricePerItem = newPrice ? newPrice : item.price;
    let newSubtotal = newQty * pricePerItem;

    // BOGO rule
    let amountToDiscountFromBOGO = 0;
    relevantRule = rules.find(rule => rule.appliesTo === type && rule.type === BOGO_STRING);
    if (relevantRule) {
      const buy = relevantRule.details.buy;
      const get = relevantRule.details.get;
      const difference = get - buy;

      const numberOfSets = Math.floor(newQty / get);
      amountToDiscountFromBOGO = difference * numberOfSets * pricePerItem;
    }
    newSubtotal = newSubtotal - amountToDiscountFromBOGO;

    return newSubtotal;
  }

  quantityChange(type: string, newQty: number) {
    const index = this.state.items.findIndex(item => item.name === type);
    const item = this.state.items[index];
    const updatedItems = [...this.state.items];

    const newSubtotal = this.applyPricingRules(item, type, newQty);

    updatedItems[index] = { ...updatedItems[index], quantity: newQty };
    updatedItems[index] = { ...updatedItems[index], subtotal: newSubtotal };

    this.setState({
      items: updatedItems,
    })
  }

  profileChange(type: string) {
    // Pricing rules change according to the profile
    const index = PRICING_RULES.findIndex(rule => rule.customerName === type);
    const updatedRules = index >= 0 ? PRICING_RULES[index].rules : [];

    const updatedItems = [...this.state.items];
    for (let i = 0; i < updatedItems.length; i++) {
      updatedItems[i] = { ...updatedItems[i], quantity: 0 };
      updatedItems[i] = { ...updatedItems[i], subtotal: 0 };
    }

    this.setState({
      profile: type,
      items: updatedItems,
      rules: updatedRules,
    })
  }

  render() {
    return (
      <Stack space="medium">
        <Box background="brand" paddingY="xxlarge" paddingX="gutter">
          <Stack space="medium">
            <Heading level="1">Your cart</Heading>
            <Text>
              <Strong>First,</Strong> select a profile.<br />
              <Strong>Then,</Strong> add or remove items and see the total change!
          </Text>
          </Stack>
        </Box>
        <Box paddingX={['xsmall', 'gutter']} style={{ marginTop: '-40px' }}>
          <Card>
            <Stack space="medium">
              <Stack space="large" dividers>
                <Profile
                  onProfileChange={this.profileChange}
                />
                <OrderSummary
                  onQuantityChange={this.quantityChange}
                  rules={this.state.rules}
                  items={this.state.items}
                />
                <GrandTotal
                  items={this.state.items}
                />
              </Stack>
            </Stack>
          </Card>
        </Box>
      </Stack>
    );
  }
}

export default Main;
export type { Item };
