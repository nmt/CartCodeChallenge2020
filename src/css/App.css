.profileSelector {
    margin: 10px 0 20px;
}

.quantityPicker,
.quantityPickerButtonHolder {
    display: flex;
}

.quantityPickerText {
    padding: 10px;
    min-width: 20px;
    text-align: right;
}

.quantityPickerText p {
    margin: 0;
}

.quantityPickerButtonHolder {
    flex-direction: column;
}

.orderSummary thead td {
    font-weight: 600;
}

.orderSummary td {
    min-width: 100px;
}

.orderSummaryItemSubtotal {
    min-width: 200px;
}

.gtLabel {
    display: inline-block;
    min-width: 200px;
    text-align: right;
    margin-right: 7px;
}
