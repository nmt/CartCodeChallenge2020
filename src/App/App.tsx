import 'braid-design-system/reset';

import { BraidLoadableProvider } from 'braid-design-system';
import React from 'react';

import Main from './Main';

interface AppProps {
  site: string;
}

export default ({ site }: AppProps) => (
  <BraidLoadableProvider themeName={site}>
    <Main />
  </BraidLoadableProvider>
);
