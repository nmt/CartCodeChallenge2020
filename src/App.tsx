import React from 'react';
import './css/App.css';
import Profile from './components/Profile';
import OrderSummary from './components/OrderSummary';
import GrandTotal from './components/GrandTotal';

type state = {
  profile: string,
  items: Array<Item>,
};

interface Item {
  name: string,
  quantity: number,
  price: number,
}

class App extends React.Component<{}, state> {
  constructor(props: Object) {
    super(props);

    this.profileChange = this.profileChange.bind(this);
    this.quantityChange = this.quantityChange.bind(this);

    this.state = {
      profile: '',
      items: [
        {
          name: 'classic',
          quantity: 0,
          price: 26999
        },
        {
          name: 'standOut',
          quantity: 0,
          price: 32299
        },
        {
          name: 'premium',
          quantity: 0,
          price: 39499
        }],
    };
  }

  quantityChange(type: string, newQty: number) {
    const index = this.state.items.findIndex(item => item.name === type);
    let newArray = [...this.state.items];
    newArray[index] = {...newArray[index], quantity: newQty}

    this.setState({
      items: newArray,
    })
  }

  profileChange(e: any) {
    this.setState({
      profile: e.target.value,
    })
  }

  render() {
    return (
      <div id="main">
        <Profile
          onProfileChange={this.profileChange}
        />
        <OrderSummary
          onQuantityChange={this.quantityChange}
          items={this.state.items}
        />
        <GrandTotal
          items={this.state.items}
        />
      </div>
    );
  }
}

export default App;
