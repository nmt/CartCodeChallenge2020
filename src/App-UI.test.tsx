import React from 'react';
import { render, screen } from '@testing-library/react';
import App from './App';

test('renders profile selector', () => {
  render(<App />);
  const linkElement = screen.getByText(/Select a profile/i);
  expect(linkElement).toBeInTheDocument();
});

test('renders order summary table', () => {
  render(<App />);
  const linkElement = screen.getByRole(/table/);
  expect(linkElement).toBeInTheDocument();
});

test('renders grand total', () => {
  render(<App />);
  const linkElement = screen.getByText(/Grand total/i);
  expect(linkElement).toBeInTheDocument();
});
