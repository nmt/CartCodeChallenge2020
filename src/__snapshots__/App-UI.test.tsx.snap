// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App snapshot 1`] = `
<div
  id="main"
>
  <div
    className="profileSelector"
    id="profileSelector"
  >
    <label
      htmlFor="profile"
    >
      Select a profile: 
    </label>
    <select
      name="profile"
      onChange={[Function]}
    >
      <option
        value=""
      >
        Default
      </option>
      <option
        value="secondBite"
      >
        SecondBite
      </option>
      <option
        value="axil"
      >
        Axil Coffee Roasters
      </option>
      <option
        value="myer"
      >
        MYER
      </option>
    </select>
  </div>
  <div
    className="orderSummary"
    id="orderSummary"
  >
    <table>
      <thead>
        <tr>
          <td>
            Name
          </td>
          <td>
            Quantity
          </td>
          <td>
            Price
          </td>
          <td>
            Total
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Classic Ad
          </td>
          <td>
            <div
              className="quantityPicker"
              id="classicAdPicker"
            >
              <div
                className="quantityPickerText"
              >
                <p>
                  0
                </p>
              </div>
              <div
                className="quantityPickerButtonHolder"
              >
                <button
                  onClick={[Function]}
                >
                  +
                </button>
                <button
                  onClick={[Function]}
                >
                  -
                </button>
              </div>
            </div>
          </td>
          <td>
            A$269.99
          </td>
          <td
            className="orderSummaryItemSubtotal"
          >
            A$0.00
          </td>
        </tr>
        <tr>
          <td>
            Stand Out Ad
          </td>
          <td>
            <div
              className="quantityPicker"
              id="standOutAdPicker"
            >
              <div
                className="quantityPickerText"
              >
                <p>
                  0
                </p>
              </div>
              <div
                className="quantityPickerButtonHolder"
              >
                <button
                  onClick={[Function]}
                >
                  +
                </button>
                <button
                  onClick={[Function]}
                >
                  -
                </button>
              </div>
            </div>
          </td>
          <td>
            A$322.99
          </td>
          <td
            className="orderSummaryItemSubtotal"
          >
            A$0.00
          </td>
        </tr>
        <tr>
          <td>
            Premium Ad
          </td>
          <td>
            <div
              className="quantityPicker"
              id="premiumAdPicker"
            >
              <div
                className="quantityPickerText"
              >
                <p>
                  0
                </p>
              </div>
              <div
                className="quantityPickerButtonHolder"
              >
                <button
                  onClick={[Function]}
                >
                  +
                </button>
                <button
                  onClick={[Function]}
                >
                  -
                </button>
              </div>
            </div>
          </td>
          <td>
            A$394.99
          </td>
          <td
            className="orderSummaryItemSubtotal"
          >
            A$0.00
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    id="grandTotal"
  >
    <p>
      <span
        className="gtLabel"
      >
        Grand total: 
      </span>
      <span
        className="rightAlign"
      >
        A$0.00
      </span>
    </p>
  </div>
</div>
`;
